service/storage_service.cc

```diff
@@ -369,7 +369,8 @@ void storage_service::prepare_to_join(
     slogger.info("Starting up server gossip");
 
     auto generation_number = db::system_keyspace::increment_and_get_generation().get0();
-    _gossiper.start_gossiping(generation_number, app_states, gms::bind_messaging_port(bool(do_bind))).get();
+    auto advertise = gms::advertise_myself(!replacing_a_node_with_same_ip);
```

Comment:
Why only in `!replacing_a_node_with_same_ip`, what was the problem ?

Reply:
1) The problem with replacing using different ip + echo message delay is that, before the existing nodes can mark the replacing node with diff ip address, existing node wont talk to the new replacing node, because it is not in either live nodes list or down node lists. As a result, the replacing node will not learn the gossip state for existing nodes.

2) The issue this patch fixes is limited to replacing_a_node_with_same_ip only. It is best to limit the fix for to it as well.

Reply:
I still don't understand. Aren't existing nodes responding to SYNs which the replacing node sends? That's what they do when the node was in the shadow round, and they don't add nodes in the shadow round to live/down lists.

Reply:
The syn from replacing node contains the node itself only. Gossip only exchanges information for the nodes listed  in the syn message. So the syn from the  replacing node will not make replacing node learn other exiting nodes.

Syn from the existing node to the replacing node (new node) will syn the existing node to the replacing node because existing nodes will contain existing nodes in syn message. However, existing nodes hasn't mark the replacing node as live. So it does not in live list or down list, in each gossip round on existing nodes, no syn message will be send to replacing node. 

Here is an example:

Node 4 is the replacing node replaces Node3 with a diff  ip address. 

![image](https://user-images.githubusercontent.com/320076/105431947-9d0fbe00-5c91-11eb-9da3-ae9347c03646.png)

Reply:
> Now I understand. So in the case of replacing with the same IP, the node will be seen by existing nodes as down, but they will still gossip with it due to do_gossip_to_unreachable_member()? It gossips with a probability which decreases with the number of live nodes in the cluster. I think we shouldn't rely on it for bootstrap, there's a chance that the node won't be contacted in large clusters. If the SYN didn't contain any digests, the existing node would respond with everything it knows (like in the shadow round). Better do that, that is send a shadow syn when bootstrapping.

The probability on a single node is unreachable_endpoint_count / (live_endpoint_count + 1),  but we have multiple live endpoints in the cluster. Any of them  talks to the replacing node is enough. So the actual probability cluster wide is live_endpoint_count * unreachable_endpoint_count / (live_endpoint_count + 1) which is very big. 

We do not need the trick to empty SYN for the fix to work. It is best to limit the scope of the fix to the minimum. 

BTW: Your reply breaks the thread in the GH page.

Reply:
> > Now I understand. So in the case of replacing with the same IP, the node will be seen by existing nodes as down, but they will still gossip with it due to do_gossip_to_unreachable_member()? It gossips with a probability which decreases with the number of live nodes in the cluster. I think we shouldn't rely on it for bootstrap, there's a chance that the node won't be contacted in large clusters. If the SYN didn't contain any digests, the existing node would respond with everything it knows (like in the shadow round). Better do that, that is send a shadow syn when bootstrapping.
> 
> The probability on a single node is unreachable_endpoint_count / (live_endpoint_count + 1), but we have multiple live endpoints in the cluster. Any of them talks to the replacing node is enough. So the actual probability cluster wide is live_endpoint_count * unreachable_endpoint_count / (live_endpoint_count + 1) which is very big.

Your formula can give larger than 1 probability. I think the probability for a single round to contact the node is actually:

`1 - (1 - unreachable_endpoint_count / (live_endpoint_count+1))^(live_endpoint_count)`

Which is about 60% for unreachable_endpoint_count = 1. It quickly approaches 100% after several rounds though, so it's probably still good enough if we wait for several rounds.

Please document the reasons for the special casing (other nodes won't contact us) in the code, and the reasons why we think they will.

Reply:
Yes,  Pall_not_talk =  (1 - Ptalk) ^ live_endpoint_count is the correct probability. 

Add more docs in v9.